File: CodingInterviews/java/PrintListFromTailToHead.java
package algorithm;

import java.util.ArrayList;

class ListNode {
	int val;
	ListNode next = null;

	ListNode(int val) {
		this.val = val;
	}
}

public class PrintListFromTailToHead {
	/* 顺序遍历链表，反转链表，最后遍历新链表并顺序获取val */
	public static ArrayList<Integer> printListFromTailToHead(ListNode listNode) {
		ListNode newHead = null;
		ListNode pNode = listNode;
		while(pNode != null){
			if(newHead == null){
				newHead = pNode;
				pNode = pNode.next;
				newHead.next = null;
			}
			else{
				ListNode tempNode = pNode;
				pNode = pNode.next;
				tempNode.next = newHead;
				newHead = tempNode;
			}
		}
		ArrayList<Integer> resList = new ArrayList<Integer>(); 
		if(newHead == null)
			return resList;
		while(newHead != null){
			resList.add(newHead.val);
			newHead = newHead.next;
		}
		return resList;
    }

	public static void main(String[] args) {
		int[] arr = { 1, 2, 3};
		ListNode preNode = null;
		ListNode head = null;
		for (int i : arr) {
			ListNode node = new ListNode(i);
			if (preNode == null) {
				preNode = node;
				head = preNode;
			} else {
				preNode.next = node;
				preNode = node;
			}

		}
		ArrayList res = printListFromTailToHead(head);
		System.out.println(res);
	}
}

File: CodingInterviews/python/print_list_from_tail_to_head.py
# -*- coding:utf-8 -*-
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

'''
递归法
'''
class Solution:
    # 返回从尾部到头部的列表值序列，例如[1,2,3]
    def printListFromTailToHead(self, listNode):
        if listNode is None:
            return []
        res = self.printListFromTailToHead(listNode.next)
        res.append(listNode.val)
        return res

if __name__ == '__main__':
    head = ListNode(0)
    pNode = head
    for i in xrange(1,7):
        tempNode = ListNode(i)
        pNode.next = tempNode
        pNode = tempNode
    solution = Solution()
    res = solution.printListFromTailToHead(head)
    print res